docs/tutorials/event_driven.md

```diff
@@ -0,0 +1,298 @@
+        .keyBy((TaxiFare fare) -> fare.driverId)
+        .process(new PseudoWindow(Time.hours(1)));
+{% endhighlight %}
+
+In this code snippet a `KeyedProcessFunction` called `PseudoWindow` is being applied to a keyed
+stream, the result of which is a `DataStream<Tuple3<Long, Long, Float>>` (the same kind of stream
+produced by the implementation that uses Flink's built-in TimeWindows).
+
+The overall outline of `PseudoWindow` has this shape:
+
+{% highlight java %}
+// Compute the sum of the tips for each driver in hour-long windows.
+// The keys are driverIds.
+public static class PseudoWindow extends 
+        KeyedProcessFunction<Long, TaxiFare, Tuple3<Long, Long, Float>> {
+
+    private final long durationMsec;
+
+    public PseudoWindow(Time duration) {
+        this.durationMsec = duration.toMilliseconds();
+    }
+
+    @Override
+    // Called once during initialization.
+    public void open(Configuration conf) {
+        . . .
+    }
+
+    @Override
+    // Called as each fare arrives to be processed.
+    public void processElement(
+            TaxiFare fare,
+            Context ctx,
+            Collector<Tuple3<Long, Long, Float>> out) throws Exception {
+
+        . . .
+    }
+
+    @Override
+    // Called when the current watermark indicates that a window is now complete.
+    public void onTimer(long timestamp, 
+            OnTimerContext context, 
+            Collector<Tuple3<Long, Long, Float>> out) throws Exception {
+
+        . . .
+    }
+}
+{% endhighlight %}
+
+Things to be aware of:
+
+* There are several types of ProcessFunctions -- this is a KeyedProcessFunction, but there are also
+  CoProcessFunctions, BroadcastProcessFunctions, etc. 
+
+* A KeyedProcessFunction is a kind of RichFunction. Being a RichFunction, it has access to the open
+  and getRuntimeContext methods needed for working with managed keyed state.
+
+* There are two callbacks to implement: `processElement` and `onTimer`. `processElement` is called
+  with each incoming event; `onTimer` is called when timers fire. These can be either event time or
+  processing time timers. Both `processElement` and `onTimer` are provided with a context object
+  that can be used to interact with a `TimerService` (among other things). Both callbacks are also
+  passed a `Collector` that can be used to emit results.
+
+#### The `open()` method
+
+{% highlight java %}
+// Keyed, managed state, with an entry for each window, keyed by the window's end time.
+// There is a separate MapState object for each driver.
+private transient MapState<Long, Float> sumOfTips;
+
+@Override
+public void open(Configuration conf) {
+
+    MapStateDescriptor<Long, Float> sumDesc =
+            new MapStateDescriptor<>("sumOfTips", Long.class, Float.class);
+    sumOfTips = getRuntimeContext().getMapState(sumDesc);
+}
+{% endhighlight %}
+
+Because the fare events can arrive out-of-order, it will sometimes be necessary to process events
```

Comment:
out-of-order event vs. event being out of order?
```suggestion
Because the fare events can arrive out of order, it will sometimes be necessary to process events
```